_posts/2019-10-23-hhvm-4.28.markdown

```diff
@@ -0,0 +1,56 @@
+---
+title: "HHVM 4.28"
+layout: post
+author: jjergus
+category: blog
+---
+
+HHVM 4.28 is released! This release marks the end of support for 4.22;
+4.23&ndash;4.27 remain supported, as do the LTS releases 3.30 and 4.8.
+
+Release 4.28.1 fixes some minor bugs that were introduced in 4.28.0.
+
+# Highlights
+
+- New implementation of auto-completion for LSP-based editors, fixing issues
+  with namespaces.
+- A new function,
+  [`set_pre_timeout_handler`](https://docs.hhvm.com/hack/reference/function/set_pre_timeout_handler/),
+  was added. The pre-timeout handler allows cheaper debugging of timeouts by
+  invoking analysis code a few seconds prior to requests timing out. It also
+  allows re-registering a new pre-timeout callback much like the memory
+  threshold callbacks.
+- Fixed a bug that would cause `hackfmt` to reformat code in a way that breaks
+  `HH_FIXME` comments in some cases.
+- It is no longer a type error to reference the built-in type
+  `HH\BuiltinEnum`&mdash;the parent type of all `enum`s (previously error 2053).
+
+# Breaking Changes
+
+- Declaring a type with the same name as a built-in type (e.g. `int`, `Vector`,
+  `Awaitable`) will no longer override the built-in type. Such types can now
+  only be referenced using their namespaced name (e.g. `MyNamespace\Vector`, or
+  `namespace\Vector` if in the same namespace).
+- It is now a type error to `use` multiple traits that declare `__construct()`
+  from the same class.
+- Trying to call a non-static method &ldquo;statically&rdquo; will now always
+  trigger a fatal error. This was already true in most cases, except some
+  special cases where it was previously only a warning (e.g. using a non-static
+  method in `array_map(['ClassName', 'nonStaticMethodName'], $arr)`).
+- It is now a type error to pass a value of type `array` where a tuple is
```

Comment:
might need to add a type-error-driven hhast migration for this :( unique code?

Reply:
Actually this is just for `array` with no generic types, I should probably make that clear. And `array` with no generic types has been an error (in strict mode) for a long time, so hopefully this isn't too common.

Reply:
Basically there's no code that would be valid in strict mode with no FIXMEs that we're breaking.